propolis/src/hw/nvme/cmds.rs

```diff
@@ -0,0 +1,560 @@
+use super::bits::{self, RawSubmission, StatusCodeType};
+use super::queue::QueueCreateErr;
+use crate::common::*;
+use crate::hw::nvme::bits::STS_INVAL_PRP_OFFSET;
+use crate::vmm::MemCtx;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum ParseErr {
+    #[error("SGLs not supported")]
+    SGL,
+
+    #[error("Fused ops not supported")]
+    Fused,
+
+    #[error("reserved PSDT value specified")]
+    ReservedPsdt,
+
+    #[error("reserved FUSE value specified")]
+    ReservedFuse,
+}
+
+type QueueId = u16;
+
+#[derive(Debug)]
+pub enum AdminCmd {
+    DeleteIOSubQ(QueueId),
+    CreateIOSubQ(CreateIOSQCmd),
+    GetLogPage(GetLogPageCmd),
+    DeleteIOCompQ(QueueId),
+    CreateIOCompQ(CreateIOCQCmd),
+    Identify(IdentifyCmd),
+    Abort,
+    SetFeatures(SetFeaturesCmd),
+    GetFeatures,
+    AsyncEventReq,
+    Unknown(RawSubmission),
+}
+impl AdminCmd {
+    pub fn parse(
+        raw: RawSubmission,
+    ) -> Result<(Self, SubmissionEntry), ParseErr> {
+        let cmd = match raw.opcode() {
+            bits::ADMIN_OPC_DELETE_IO_SQ => {
+                AdminCmd::DeleteIOSubQ(raw.cdw10 as u16)
+            }
+            bits::ADMIN_OPC_CREATE_IO_SQ => {
+                let queue_prio = match (raw.cdw11 & 0b110) >> 1 {
+                    0b00 => QueuePriority::Urgent,
+                    0b01 => QueuePriority::High,
+                    0b10 => QueuePriority::Medium,
+                    0b11 => QueuePriority::Low,
+                    _ => unreachable!(),
+                };
+                AdminCmd::CreateIOSubQ(CreateIOSQCmd {
+                    prp: raw.prp1,
+                    qsize: (raw.cdw10 >> 16) as u16,
+                    qid: raw.cdw10 as u16,
+                    cqid: (raw.cdw11 >> 16) as u16,
+                    queue_prio,
+                    phys_contig: (raw.cdw11 & 1) != 0,
+                })
+            }
+            bits::ADMIN_OPC_GET_LOG_PAGE => {
+                AdminCmd::GetLogPage(GetLogPageCmd {
+                    nsid: raw.nsid,
+                    len: (((raw.cdw11 as u16) as u32) << 16
+                        | (raw.cdw10 >> 16))
+                        * 4,
```

Comment:
In NVMe 1.0 only bits 27:16 of cdw10 are defined to contain a length. In general we're advertising 1.0 but we have a lot of stuff we're parsing for later revisions. Not sure whether or not this matters right at this phase, but eventually we'll want to make sure we're only checking and honoring bits based on whatever version we actually define.

Reply:
Yup, cleaned up the mismatch in advertised version vs what the code tried to parse/do.